apiextensions: validate status.observedGeneration for custom resources

nikhita opened the pull request:
Fixes https://github.com/kubernetes/features/issues/95#issuecomment-302946176
Fixes https://github.com/kubernetes/kubernetes/issues/7328#issuecomment-359570791, https://github.com/kubernetes/kubernetes/issues/7328#issuecomment-305772161

We already increment `metadata.Generation` on spec updates and users can technically have  a `status.observedGeneration`. The external controller is responsible for updating the value but we should still be validating it.

This PR adds the following validations for `status.observedGeneration`:

1. it should be a non-negative integer.
2. it should never be greater than the `metadata.generation` value.
3. it should never be greater than the previously observed `status.observedGeneration` value.

/cc sttts deads2k 
/assign sttts 
/sig api-machinery
/area custom-resources

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
For custom resources, the `.status.observedGeneration` value is now validated to ensure that it is a non-negative integer and that it is smaller than the `metadata.generation` value and the previously observed `status.observedGeneration` value.
```

nikhita commented:
@kargakis @ash2k since you had asked for it in https://github.com/kubernetes/features/issues/95#issuecomment-302946176.

nikhita commented:
We _can_ also support `status.ObservedGeneration` for CRDs but I can't think of places where that would make sense.

ash2k commented:
@nikhita Hey, thanks a lot for implementing this and for the heads up. 

Is it valid to decrement `status.observedGeneration`? It should only be increasing right? Also, it should not be greater than `metadata.Generation`, right?

sttts commented:
@ash2k this sounds reasonable. But I would like to hear @soltysh @kargakis to shed some light on the intended semantics, not that we miss anything.

nikhita commented:
> Is it valid to decrement status.observedGeneration? It should only be increasing right? Also, it should not be greater than metadata.Generation, right?

I agree. Generation number is always monotonically increasing. The generation number observed by the controller i.e. `observedGeneration` should never be greater than the generation number the apiserver has processed. (honestly, kind of scary :scream:)

However, I notice that all other resources that support `observedGeneration` allow `observedGeneration > generation`. Example:

https://github.com/kubernetes/kubernetes/blob/b9e46f542266d6c3ba7d76731cfa8c707bb7e456/pkg/kubectl/rollout_status.go#L80

@kargakis @soltysh Is there a reason why we allow this? Shouldn't this just be `observedGeneration == generation`?

soltysh commented:
> @kargakis @soltysh Is there a reason why we allow this? Shouldn't this just be observedGeneration == generation?

@janetkuo or @tnozicka might shed some light on `observedGeneration` as well, I haven't implemented that mechanism in deployments but will need to do in jobs, most probably.

kargakis commented:
Generation is increased by the api server on spec changes, observed
generation is meant to be set to generation by the controller when it
processes the spec (change). This is helpful for clients to track whether
their changes have been observed by the controller or to verify that they
are looking into an up to date resource (metadata.generation ==
status.observedgeneration).

On Mon, May 28, 2018, 00:37 Dr. Stefan Schimanski <notifications@github.com>
wrote:

> @ash2k <https://github.com/ash2k> this sounds reasonable. But I would
> like to hear @soltysh <https://github.com/soltysh> @kargakis
> <https://github.com/kargakis> to shed some light on the intended
> semantics, not that we miss anything.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/64379#issuecomment-392445293>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADuFf68UYew-cOJAPYiLBj6nnnyLqlxFks5t26klgaJpZM4UPaPA>
> .
>

kargakis commented:
For the validation part I don't remember why we allow greater observed
generation, maybe there is a scenario in the future where spec and status
get separated and you may get a stale generation back due to multiple
running api servers? We wouldn't want to invalidate those resources that's
why I think we left it out.

On Mon, May 28, 2018, 00:44 Nikhita Raghunath <notifications@github.com>
wrote:

> Is it valid to decrement status.observedGeneration? It should only be
> increasing right? Also, it should not be greater than metadata.Generation,
> right?
>
> I agree. Generation number is always monotonically increasing. The
> generation number observed by the controller i.e. observedGeneration
> should never be greater than the generation number the apiserver has
> processed. (honestly, kind of scary 😱)
>
> However, I notice that all other resources that support observedGeneration
> allow observedGeneration > generation. Example:
>
>
> https://github.com/kubernetes/kubernetes/blob/b9e46f542266d6c3ba7d76731cfa8c707bb7e456/pkg/kubectl/rollout_status.go#L80
>
> @kargakis <https://github.com/kargakis> @soltysh
> <https://github.com/soltysh> Is there a reason why we allow this?
> Shouldn't this just be observedGeneration == generation?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/64379#issuecomment-392446869>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADuFf_Eer_lOeeLN1UntYv3nN5n0nn79ks5t26rvgaJpZM4UPaPA>
> .
>

fedebongio commented:
/cc @mbohlool

nikhita commented:
> maybe there is a scenario in the future where spec and status get separated and you may get a stale generation back due to multiple running api servers

@kargakis Can you elaborate on how the spec and the status would get separated? Cannot think of a scenario where this would be possible.

nikhita commented:
bumping this again

@kargakis can you elaborate?

https://github.com/kubernetes/kubernetes/blob/27c327cd33db8c91f1c6ad6c701b0457c780bce2/pkg/kubectl/rollout_status.go#L80-L87

treats it as observedGeneration == generation even though the if condition says otherwise.

kargakis commented:
maybe @janetkuo @bgrant0607 remember why we did it

enisoc commented:
@nikhita wrote:
> Can you elaborate on how the spec and the status would get separated?

I can't find the issue right now, but I recall seeing an old discussion about separating spec and status into separate resources, which could even potentially be stored in separate etcd instances like events. If that were to happen, it would become possible for a controller to see a status that's more up-to-date than the spec it sees.

I haven't seen any recent discussion of that, so it seems unlikely to happen any time soon. As long as spec and status are stored together in the same object, I can't think of a way for ObservedGeneration to be greater than Generation except for a user error or programming error.

nikhita committed:
apiextensions: validate status.observedGeneration for custom resources

`status.observedGeneration` is useful to identify which `generation`
the status is valid for. This is mostly applicable when multiple
controllers are acting on the same object.

Right now, we already increment `metadata.Generation` on spec updates
and users can have  a `status.ObservedGeneration`.
However, we do not validate it. This commit adds the following
validations for it:

1. it should be a non-negative integer.
2. it should never be greater than the metadata.generation value.
3. it should never be greater than the previously observed
status.observedGeneration value.

nikhita commented:
Updated according to discussions in this PR + in https://github.com/kubernetes/kubernetes/pull/66501.

Now, this PR adds the following validations for `status.observedGeneration`:

1. it should be a non-negative integer.
2. it should never be greater than the `metadata.generation` value.
3. it should never be greater than the previously observed `status.observedGeneration` value.

Ptal, thanks!

nikhita commented:
/cc @liggitt

nikhita commented:
/retest

On Tue, Aug 21, 2018 at 4:22 PM k8s-ci-robot <notifications@github.com>
wrote:

> @nikhita <https://github.com/nikhita>: The following test *failed*, say
> /retest to rerun them all:
> Test name Commit Details Rerun command
> pull-kubernetes-e2e-gce-100-performance 6332171
> <https://github.com/kubernetes/kubernetes/commit/633217188350502daa68875332607dd8cb7b6f4c>
> link
> <https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/64379/pull-kubernetes-e2e-gce-100-performance/17115/> /test
> pull-kubernetes-e2e-gce-100-performance
>
> Full PR test history <https://k8s-gubernator.appspot.com/pr/64379>. Your
> PR dashboard <https://k8s-gubernator.appspot.com/pr/nikhita>. Please help
> us cut down on flakes by linking to
> <https://git.k8s.io/community/contributors/devel/flaky-tests.md#filing-issues-for-flaky-tests>
> an open issue
> <https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open> when
> you hit one in your PR.
>
> Instructions for interacting with me using PR comments are available here
> <https://git.k8s.io/community/contributors/guide/pull-requests.md>. If
> you have questions or suggestions related to my behavior, please file an
> issue against the kubernetes/test-infra
> <https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:>
> repository. I understand the commands that are listed here
> <https://go.k8s.io/bot-commands>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/64379#issuecomment-414633634>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APXA0C5Cajanpb2JBiis3jHuEQgjmQ_Sks5uS-ZWgaJpZM4UPaPA>
> .
>

nikhita commented:
/retest

On Tue, Aug 21, 2018 at 5:26 PM k8s-ci-robot <notifications@github.com>
wrote:

> @nikhita <https://github.com/nikhita>: The following test *failed*, say
> /retest to rerun them all:
> Test name Commit Details Rerun command
> pull-kubernetes-e2e-gce-100-performance 6332171
> <https://github.com/kubernetes/kubernetes/commit/633217188350502daa68875332607dd8cb7b6f4c>
> link
> <https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/64379/pull-kubernetes-e2e-gce-100-performance/17128/> /test
> pull-kubernetes-e2e-gce-100-performance
>
> Full PR test history <https://k8s-gubernator.appspot.com/pr/64379>. Your
> PR dashboard <https://k8s-gubernator.appspot.com/pr/nikhita>. Please help
> us cut down on flakes by linking to
> <https://git.k8s.io/community/contributors/devel/flaky-tests.md#filing-issues-for-flaky-tests>
> an open issue
> <https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open> when
> you hit one in your PR.
>
> Instructions for interacting with me using PR comments are available here
> <https://git.k8s.io/community/contributors/guide/pull-requests.md>. If
> you have questions or suggestions related to my behavior, please file an
> issue against the kubernetes/test-infra
> <https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:>
> repository. I understand the commands that are listed here
> <https://go.k8s.io/bot-commands>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/64379#issuecomment-414648293>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APXA0MV_txXMIR9BLcDWruGDh279Wk_gks5uS_VagaJpZM4UPaPA>
> .
>

liggitt commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go:
```diff
@@ -397,3 +397,19 @@ func (u *Unstructured) GetClusterName() string {
 }
+
+func (u *Unstructured) GetObservedGeneration() int64 {
+	val, found, err := NestedInt64(u.Object, "status", "observedGeneration")
```
I don't think this belongs as a first-class method... if custom resources want to do this, they can call `NestedInt64(u.Object, "status", "observedGeneration")`

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go:
```diff
@@ -193,3 +200,36 @@ func (a customResourceValidator) ValidateScaleStatus(ctx context.Context, obj *u
+
+	observedGeneration, exists, err := unstructured.NestedInt64(obj.UnstructuredContent(), "status", "observedGeneration")
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("status.observedGeneration"), observedGeneration, err.Error()))
```
how does this deal with existing persisted data, since this is tightening validation?

yue9944882 commented:
/cc

ash2k commented:
/remove-lifecycle stale

bgrant0607 commented:
Reminder: assume I don't see github notifications in realtime and that searching for mentions doesn't work reliably. I'm just now noticing this.

First, regarding validation: In order to enforce invariants on observedGeneration, a controller would have to use resourceVersion to ensure that it's not acting on an old version of the spec. If it's not and it's just writing the status, then the value could well be less than the most recent values of generation and/or observedGeneration. I'm not sure that validation is the best way to prevent that, since it wouldn't necessarily prevent the controller from doing the wrong thing, only from reporting it. We also need to think about disaster recovery scenarios.

Re. the third point: "should never be greater than the previously observed status.observedGeneration" -- I assume that's backwards if we wanted observedGeneration to be monotonically increasing.

Re. separate spec and status: #8625. But I'm not sure that has anything to do with this particular issue. (Tip: #8190 lists a lot of long-standing API-related issues)

The kubectl generation/observedGeneration check was added by #19946. I don't think I reviewed it. My guess is that the `<=` was a result of reordering the code. If generation were greater than observedGeneration, then the function could have returned early.

liggitt commented:
I don't see a lot of value in this validation on the server side. This is part of the bookkeeping the controller does in the status to know if it has reacted or needs to react to the metadata.generation. Starting to validate specific subfields of status seems like protecting the controller from itself.

nikhita commented:
> Starting to validate specific subfields of status seems like protecting the controller from itself.

:+1: 

/close
